desc/optimize/_constraint_wrappers.py

```diff
@@ -1116,60 +1120,96 @@ def jvp_unscaled(self, v, x, constants=None):
         xg, xf = self._update_equilibrium(x, store=True)
-        jvpfun = lambda u: self._jvp(u, xf, xg, constants, op="unscaled")
-        return batched_vectorize(
-            jvpfun, signature="(n)->(k)", chunk_size=self._objective._jac_chunk_size
+
+        # we don't need to divide this part into blocked and batched because
+        # self._constraint._deriv_mode will handle it
+        jvpfun = lambda u: self._get_tangent(u, xf, constants, op=op)
+        tangents = batched_vectorize(
+            jvpfun,
+            signature="(n)->(k)",
+            chunk_size=self._constraint._jac_chunk_size,
         )(v)
 
-    def _jvp(self, v, xf, xg, constants, op):
-        # we're replacing stuff like this with jvps
-        # Fx_reduced = Fx[:, unfixed_idx] @ Z               # noqa: E800
-        # Gx_reduced = Gx[:, unfixed_idx] @ Z               # noqa: E800
-        # Fc = Fx @ dxdc @ v                                # noqa: E800
-        # Gc = Gx @ dxdc @ v                                # noqa: E800
-        # LHS = Gx_reduced @ (Fx_reduced_inv @ Fc) - Gc     # noqa: E800
+        if self._objective._deriv_mode == "batched":
+            # objective's method already know about its jac_chunk_size
+            return getattr(self._objective, "jvp_" + op)(tangents, xg, constants[0])
+        else:
+            return _proximal_jvp_blocked_pure(
+                self._objective,
+                jnp.split(tangents, np.cumsum(self._dimx_per_thing), axis=-1),
+                jnp.split(xg, np.cumsum(self._dimx_per_thing)),
+                op,
+            )
+
+    def _get_tangent(self, v, xf, constants, op):
+        # Note: This function is vectorized over v. So, v is expected to be 1D array
+        # of size self.dim_x.
 
-        # v contains "boundary" dofs from eq and other objects
-        # want jvp_f to only get parts from equilibrium, not other things
+        # v contains "boundary" dofs from eq and other objects (like coils, surfaces
+        # etc) want jvp_f to only get parts from equilibrium, not other things
         vs = jnp.split(v, np.cumsum(self._dimc_per_thing))
-        # this is Fx_reduced_inv @ Fc
+        # This is (dF/dx)^-1 * dF/dc  # noqa : E800
         dfdc = _proximal_jvp_f_pure(
             self._constraint,
             xf,
             constants[1],
             vs[self._eq_idx],
-            self._unfixed_idx,
-            self._Z,
-            self._D,
+            self._eq_solve_objective._unfixed_idx_mat,
             self._dxdc,
             op,
         )
         # broadcasting against multiple things
         dfdcs = [jnp.zeros(dim) for dim in self._dimc_per_thing]
         dfdcs[self._eq_idx] = dfdc
+        # note that size.dfdc != size.vs[self._eq_idx]
+        # so, the above line is not a simple assignment where you just change the value
+        # of the array, but we actually change the size of it too
+        # the size change will be handled by multiplication by the unfixed_idx_mat later
```

Comment:
can you explain this? really it is that `dfdc.size` is NOT the same as `self.dimc_per_thing[self._eq_idx]` right?

But I am not sure why, is this bc it also has like Ra_n and Za_n? then the unfixed idx mat will remove those?